reactive/kotlinx-coroutines-reactive/src/Publish.kt

```diff
@@ -176,17 +176,32 @@ private class PublisherCoroutine<in T>(
                 if (cancelled) {
                     // If the parent had failed to handle our exception, then we must not lose this exception
                     if (cause != null && !handled) handleCoroutineException(context, cause)
-                } else {
-                    try {
-                        if (cause != null && cause !is CancellationException) {
+                    return
+                }
+
+                try {
+                    if (cause != null && cause !is CancellationException) {
+                        /*
+                         * Reactive frameworks have two types of exceptions: regular and fatal.
+                         * Regular are passed to onError.
+                         * Fatal can be passed to onError, but implementation **is free to swallow it** (e.g. see #1297).
+                         * Such behaviour is inconsistent, leads to silent failures and we can't possibly know whether
+                         * the cause will be handled by onError (and moreover, it depends on whether a fatal exception was
+                         * thrown by subscriber or upstream).
+                         * To make behaviour consistent and least surprising, we always handle fatal exceptions
+                         * by coroutines machinery, anyway, they should not be present in regular program flow,
+                         * thus our goal here is just to expose it as soon as possible.
+                         */
+                        if (cause.isFatal()) {
```

Comment:
I'm worried that this `if` makes our implementation non compliant with the spec as we are not even trying to deliver those exceptions to the `subscriber.onError` as specification requires us to do. We don't terminate it in any way! We should first always _attempt_ to deliver the error to `subscriber.onError(cause)` and then, if the error if fatal, knowing that some reactive specs swallow then do `if (!handled && cause.isFatal()) handleCoroutineException(context, cause)` to make sure it is not lost.

Reply:
Makes sense